### DISCUSSION TITLE
Custom async IAsyncEnumerable type?

### DISCUSSION THREAD

**Discussion by timcassell** [dotnet/csharplang#4798] [Q&A] [answered]
C# 7.2 introduced builders for custom task-like types, then C# 8 introduced async enumerators. I'm wondering if it's possible to back an `async IAsyncEnumerable<T>` function with a custom type in a similar manner, and if so, how?

If I create a type that implements `IAsyncEnumerable`, `await foreach` works to consume it. Can that then be used as the return type for the async iterator function with another special builder?

```cs
public struct MyAsyncEnumerable<T> : IAsyncEnumerable<T>
{
    ...
}
```

```cs
public async MyAsyncEnumerable<int> EnumerateMyValuesAsync()
{
    await Task.Yield();
    yield return 42;
}
```

**Comment by timcassell** [comment]
@stephentoub I saw you wrote the [Microsoft article about async enumerables in November 2019](https://docs.microsoft.com/en-us/archive/msdn-magazine/2019/november/csharp-iterating-with-async-enumerables-in-csharp-8), do you have the answer to this?

**Comment by stephentoub** [comment, marked as answer]
> I'm wondering if it's possible to back an async IAsyncEnumerable<T> function with a custom type in a similar manner

It's not.

> Can that then be used as the return type for the async iterator function with another special builder?

No.  There's no ability in the language/compiler to substitute an implementation for any iterator, whether sync or async.

**Reply by neuecc** [reply]
@stephentoub
I was proposed this at #3629, 
I'm not really convinced that they made async impossible because sync is not possible.
